docs/cross-platform.md

```diff
@@ -0,0 +1,119 @@
+
+## Platform Differences
+
+### Path considerations
+
+A major difference with Windows vs OSX/Linux is the path separator - `\` vs `/`. Windows can actually work with _either_ path separator (for the most part). 
+
+The challenge, though, is `\` is also the __escape character__ - so we can hit double-escaping issues when we render the strings.
+
+For the time being, we'll __normalize all paths to use a forward slash__ - this keeps us from hitting the 'double-escaping issues' with backslash.
+
+One other complexity is that we use Cygwin for running our build tasks, and the canonical paths for Cygwin are POSIX style, like:
+- `C:\temp` -> `/cygdrive/c/temp`
+
+The majority of Cygwin utilities can work with either style of paths - therefore, we will prefer the __Windows path, normalized with a forward slash__, for shell invocations.
+
+Some utilities that Cygwin ships with do not support any style of Windows path (either forward or back slash). For those utilities - like `rsync` - we will need to convert the path to a cygwin-style path. Cygwin includes the `cygpath` utility for this, which can convert paths of the form `C:/temp` or `C:\\temp` to cygwin-style paths like `/cygdrive/c/temp`. These exceptions should be few, and primarily encountered in bootstrapping the `esy` environment internally, as opposed to present in the build scripts.
+
+### Cross-platform environment variables
+
+Another difference in Windows vs OSX/Linux is how the environment variables separate individual paths. On POSIX systems, it's a colon - `:` - on Windows, its `;`
+
+Because our __built artifacts are native__, we must use native Windows paths in our environment variables (normalized with forward slash).
+
+### Symlinks
+
+Symlinks on Windows are not well-supported today. To create symlinks, a Windows user either needs to be running as _administrator_, or set up a special developer mode - neither which is the default behavior on the system.
+
+Therefore, at this time, we should avoid using symlinks in cross-platform code.
+
+### File Permission Model
+
+The file permission models between Windows and OSX/Linux are significantly different - the OCaml API attempts to emulate some of the behavior of `stat` and `chmod`, but use caution when getting / setting file permissions in cross-platform code.
+
+### PATH limit
+
+- Windows has a `PATH` limit of `260`. Newer APIs no longer have this limit, but they are opt-in. You can also set a registry key to fix this, but again, it's opt-in. Something to be mindful of if your tool creates lots of nested paths!
+
+## Long-term plans
+
+As `esy` matures, we can minimize our dependency on `cygwin` for Windows - we can use `esy`'s DSL for specifying build commands to directly target native-windows executables, bypassing the `cygwin` layer. In parallel, we can start to migrate packages to more fully support cross-platform builds.
```

Comment:
It would be cool to allow pure-Dune projects or inherently cross platform packages to opt out. There's no reason they even need cygwin at all!

Reply:
Ah yes, for sure. One thing we discussed on Discord a bit back was making `esy-bash` a dependency (and renaming it `esy-essentials` or `esy-native-toolchain` or something). I logged an issue to track this further: #643

We could model this as an implicit dependency we add by default, but then if the user 'opts out', we wouldn't include this dependency.